Implement syncprogress callback

Daksh14 opened the pull request.

Daksh14 committed:
Implement syncprogress callback

ZER0 reviewed (changes requested):
1. Do not ask for review yet, if the Ci is not even passing: fix the code / tests first.
2. This is not what I suggested in issue where I discussed the implementation. The `onprogress` function was introduced exactly to avoid to pass too much arguments to `getOwnedNotes`, and to keep the responsibility on handling _block heights_ and _last positions_ to the caller; what `getOwnedNotes` should notify is just the **progression**, of the parsing chunks, and that can be easily done by checking the total bytes and bytes parsed.
Unless there is a valid reason to do differently, avoiding to pass the and handle the last pos and blockheight would be better.

Also, since you're passing the entire buffer to the function, you could have it as async generator, but that's just a matter of API, although it might makes sense.

Daksh14 committed:
Fix tests

ZER0 commented on example/index.html:
```diff
@@ -59,9 +59,11 @@
+                    from: 0,
+                };
+
+                let sync = await wallet.sync(syncOptions);
```
I don't quite understand why the example was changed, before it was perfectly fine.

ZER0 commented on src/crypto.js:
```diff
@@ -22,19 +22,73 @@ export function getNullifiers(wasm, [...seed], [...notes]) {
-  args.set(leaves, seed.length);
-  return call_raw(wasm, args, "check_note_ownership").then(parseEncodedJSON);
+export async function getOwnedNotes(wasm, seed, leaves, onprogress) {
+  let noteData = {
```
This can be `const` (JS `const` has a different meaning than Rust ones).

ZER0 commented on src/crypto.js:
```diff
@@ -22,19 +22,73 @@ export function getNullifiers(wasm, [...seed], [...notes]) {
+      i * bytesPerFunction,
+      (i + 1) * bytesPerFunction,
+    );
+
```
Given the documentation, `leaves` is a `Uint8Array` (and if it isn't, it should be). That means you shouldn't use `slice` but [subarray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/subarray) instead.

The first returns a (shallow) copy of the elements, while the second does not copy anything, just return a new `Uint8Array` that points to the same `ArrayBuffer`.
Since we never have overlapping chunks there is no problem here in terms of concurrency, even if we'll end up to have multiple workers.

It's basically the same of having multiple reference over a vector / array in Rust on different sections.

ZER0 commented on src/crypto.js:
```diff
@@ -22,19 +22,73 @@ export function getNullifiers(wasm, [...seed], [...notes]) {
+      parseEncodedJSON,
+    );
+
+    owned.notes.forEach((v) => noteData.notes.push(v));
```
Just use `push` with the spread operator:
```js
noteData.notes.push(...owned.notes);
noteData.pks.push(...owned.public_spend_keys);
noteData.nullifiers.push(...owned.nullifiers);
noteData.blockHeights.push(...owned.block_heights.split(",").map(Number));
noteData.lastPos = owned.lastPos;
```

ZER0 replied:
Follow-up after our call. Although we said that it shouldn't be any issue with the spread operator, since the number of elements will never exceed 100, if you want to be sure we won't have this issue we can also use the following approach:

```js
noteData.notes = [...noteData.notes, ...owned.notes];
```
or:
```js
noteData.notes = noteData.notes.concat(owned.notes);
```

ZER0 commented on src/crypto.js:
```diff
@@ -22,19 +22,73 @@ export function getNullifiers(wasm, [...seed], [...notes]) {
+
+    noteData.lastPos = owned.last_pos;
+
+    onprogress(noteData.lastPos);
```
Check the specs on the issue: `onprogress`  does not receive the `lastPos` directly, it receives the actual progression (a floating number between 0 and 1).

Also, if `onprogress` is undefined, you can avoid to call it, since we will have just one single iteration and there is no one to notify about it.

ZER0 commented on src/node.js:
```diff
@@ -99,13 +107,22 @@ export async function sync(wasm, seed, options = {}, node = NODE) {
+  if (typeof onblock !== "function") {
+    onblock = () => {};
+  }
+
```
Remove this.

ZER0 commented on src/node.js:
```diff
@@ -170,20 +180,17 @@ export async function sync(wasm, seed, options = {}, node = NODE) {
+
+    onblock(currentEstimatedBlockHeight, networkBlockHeight);
+  };
+
```
1. As mentioned, check the specs: `onprogress` doesn't take the `currentPos`.
2. Here you can check if `onblock` is a function, and if it's not, you can just set `onprogress` to `undefined`. No need to call `onprogress`, doing all the calculation, then an empty `onblock` if the consumer is not interested.

ZER0 commented on src/crypto.js:
```diff
@@ -22,19 +22,73 @@ export function getNullifiers(wasm, [...seed], [...notes]) {
  * @param {Uint8Array} seed - Seed of the wallet
  * @param {Uint8Array} leaves - leafs we get from the node
- * @returns {object} - object.last_pos, object.owned_notes
+ * @param {syncProgress} onprogress - callback for progress report
```
onprogress can be also `undefined`.

ZER0 commented on src/crypto.js:
```diff
@@ -22,19 +22,73 @@ export function getNullifiers(wasm, [...seed], [...notes]) {
+  };
+
+  const bytesPerFunction = 632 * 100;
+  const total = leaves.length / bytesPerFunction;
```
1. You need to ensure that `bytesPerFunction` is not greater than `leaves.length`. For example, assuming you're syncing 10 notes, then the `total` wouldn't make sense. The `total` here is the number of chunk, is that correct? Therefore it should be always at least `1`. 

2. You also need to manage the scenario where `onprogress` is `undefined`. In this case, `bytesPerFunction` is equal to `leaves.length`, so the `total` will be `1`, we will have one single chunk that have everything.

So, with both points in mind, you will have:

```js
const bytesPerFunction = onprogress ? Math.min(leaves.length, 632 * 100) : leaves.length;
const total = leaves.length / bytesPerFunction;
```

But I would put `632 * 100` as uppercase constant somewhere in the module, something like `LEAVES_CHUNK_SIZE` or similar.

ZER0 commented on tests/balance_test.js:
```diff
@@ -355,3 +355,22 @@ Deno.test({
+
+    await wallet.sync(syncOptions);
+  },
+});
```
You should tests also the non happy path (e.g. when the data you sync is less then `bytesPerFunction`, or you don't pass `onblock`).

Also, you don't need the arrow functions there:
```js
let syncOptions = {
      from: 0,
      onblock(current, final) {
        assertEquals(final, networkBlockHeight);
        assertEquals(typeof current, "number");
      },
    };
```

Daksh14 replied:
We already test when we dont pass the `onblock`. The data we sync in the CI test is less than `bytesPerFunction` already

ZER0 replied:
Not sure we're doing that anymore. My suggestion is, if you implement the `setItemsPerChunk`, adding a test where you set a bigger number than the actual notes you have in the `notes.rkyv`, and check that the `onblock` is called only once.

Also, you should add a test where you don't pass `onblock` for the sync but everything still works as expected (if you have previous tests for sync that already do that, you can avoid to repeat, e.g. sync and check you have the correct notes).

Also, please move `notes.rkyv` under `tests`.

Daksh14 committed:
Fix onprogress callback

Daksh14 committed:
onprogress is optional

ZER0 commented on src/crypto.js:
```diff
@@ -6,6 +6,9 @@
 import { call, call_raw } from "./wasm.js";
 import { parseEncodedJSON } from "./encoding.js";
+
+const chunkSize = 632 * 100;
```
This should be `CHUNK_SIZE`

ZER0 commented on example/index.html:
```diff
@@ -60,7 +60,10 @@
+                    from: 0,
+                    onblock: (current, final) => {
+                        console.log("b", current, final);
+                    },
```
You don't need arrow function, I suggest you to write something like:
```js
let sync = await wallet.sync({
     from: 0,
     onblock(current, total) {
          console.log(`Processing ${current} / ${final}`)
     }
})

ZER0 commented on src/crypto.js:
```diff
@@ -16,25 +19,86 @@ import { parseEncodedJSON } from "./encoding.js";
+      bytesProcessed = Math.min(bytesProcessed + chunkSize, totalBytes);
+      onprogress(bytesProcessed / totalBytes);
+    }
+  }
```
I know it's a bit counter intuitive, but we need to move this before the actual web assembly call.

Why is that: the idea is that we're notifying the consumer (the UI) about what's blocks we're going to process, not when it's been already processed. The reason is that, in this way, the UI can – if want to – prepare in advance some animation. The consumer UI also knows when the previous chunk was indeed finished to be processed, since another call to `onblock` for the next chunk will be executed. Or, the `sync` will return, so the consumer if it has some animation in progress, can deal with them.

ZER0 commented on tests/balance_test.js:
```diff
@@ -355,3 +355,22 @@ Deno.test({
+
+    let syncOptions = {
+      from: 0,
+      onblock: (current, final) => {
```
Same here, no need of arrow function.

ZER0 commented on tests/balance_test.js:
```diff
@@ -355,3 +355,22 @@ Deno.test({
 });
+
+Deno.test({
+  name: "syncprogress test",
```
This test actually does not test anything.
As mentioned you should have some mocked data that you can actually parse and chuck, and verify that everything works as expected. E.g. the number of chunks, the number of  time `onblock` is called, the args, etc. And you need some "unhappy path" as well, such as the mocked data being less than `bytesPerFunction` and not passing a `onblock` at all.

ZER0 reviewed (changes requested):
There are two big changes here to do, one is related to when call the `onprogress` (see my comment) and the other is about the tests.

You also ignored most of my previous comments on this PR.
Please apply those as well.

Daksh14 committed:
Call onprogress before processing, minor fixes

Daksh14 committed:
Use subarray instead

Daksh14 committed:
Add logic in test for chunksize notes

Daksh14 committed:
use globalThis

Daksh14 committed:
Use pathname instead of full URL

Daksh14 committed:
rename fetch

Daksh14 committed:
Move reset call before syncprogress

ZER0 commented on src/crypto.js:
```diff
@@ -6,6 +6,9 @@
 import { call, call_raw } from "./wasm.js";
 import { parseEncodedJSON } from "./encoding.js";
+
+const CHUNK_SIZE = 632 * 100;
```
I still believe the `100` part should be somehow configurable. It will helps to fine tuning the chunk size for the consumer.

Something simple atm not complex, so either a env variable, or a public method to set it for the whole module. Something like:

```js
const ITEM_SIZE = 632;
let chunkSize = ITEM_SIZE * 100;

export const setItemsPerChunk = (value) => {
   chunkSize = ITEM_SIZE * value;
}
```

Daksh14 committed:
Apply prettier, add sanitize ops property to test

ZER0 commented on src/node.js:
```diff
@@ -364,12 +368,12 @@ export async function blockHeightToLastPos(
   }
 
-  const { last_pos } = await getOwnedNotes(wasm, seed, firstNote);
+  const { lastPos } = await getOwnedNotes(wasm, seed, firstNote, () => {});
```
You shouldn't pass anything here: `onprogress` can be `undefined`, that was the whole purpose of having it not specified all the times.

Daksh14 committed:
replace foreach and push with concat

Daksh14 committed:
Fix typo

ZER0 reviewed (approved):
LGTM with the comments addressed before merging.

Daksh14 committed:
Add test, move notes file

Daksh14 committed:
fix tests

ZER0 committed:
Fix Unit Tests (unspentSpentNotes failure)

- Remove unnecessary `seedBytes` typed array
- Add `mock` module under test to group mocked behavior
- Remove unnecessary callback hell and pyramid of doom from test
- Remove global `networkBlockHeight` between tests definition
- Add `mock.withMockedFetch` to `sync` tests with `onblock` callback
- Remove unnecessary asset for encoded notes and truncate it from code
- Change the mock fetch to returns mocked data only in presence of `Rusk-Feeder` header
